packages/opencensus-exporter-trace-stackdriver/src/stackdriver.ts

```diff
@@ -0,0 +1,167 @@
+  resource: {traces: {}};
+  auth: JWT;
+}
+
+/** Format and sends span information to Stackdriver */
+export class StackdriverTraceExporter implements types.Exporter {
+  projectId: string;
+  exporterBuffer: classes.ExporterBuffer;
+  logger: types.Logger;
+  failBuffer: types.SpanContext[] = [];
+
+  constructor(options: StackdriverExporterOptions) {
+    this.projectId = options.projectId;
+    this.logger = options.logger || logger.logger();
+    this.exporterBuffer = new classes.ExporterBuffer(this, options);
+  }
+
+  /**
+   * Is called whenever a span is ended.
+   * @param root the ended span
+   */
+  onEndSpan(root: types.RootSpan) {
+    this.exporterBuffer.addToBuffer(root);
+  }
+
+  /**
+   * Publishes a list of root spans to Stackdriver.
+   * @param rootSpans
+   */
+  publish(rootSpans: types.RootSpan[]) {
+    const stackdriverTraces =
+        rootSpans.map(trace => this.translateTrace(trace));
+    return this.authorize(stackdriverTraces)
+        .then((result: TracesWithCredentials) => {
+          return this.sendTrace(result);
+        })
+        .catch(err => {
+          for (const root of rootSpans) {
+            this.failBuffer.push(root.spanContext);
+          }
+          return `${err}`;
+        });
+  }
+
+  /**
+   * Translates root span data to Stackdriver's trace format.
+   * @param root
+   */
+  private translateTrace(root: types.RootSpan) {
+    const spanList = root.spans.map(span => this.translateSpan(span));
+    spanList.push(this.translateSpan(root));
+
+    return {
+      'projectId': this.projectId,
+      'traceId': root.traceId,
+      'spans': spanList
+    };
+  }
+
+  /**
+   * Translates span data to Stackdriver's span format.
+   * @param span
+   */
+  private translateSpan(span: types.Span) {
+    return {
+      'name': span.name,
+      'kind': 'SPAN_KIND_UNSPECIFIED',
+      'spanId': span.id,
+      'startTime': span.startTime,
+      'endTime': span.endTime
+    };
+  }
+
+  /**
+   * Sends traces in the Stackdriver format to the service.
+   * @param projectId
+   * @param authClient
+   * @param stackdriverTraces
+   */
+  private sendTrace(traces: TracesWithCredentials) {
+    return new Promise((resolve, reject) => {
+      cloudTrace.projects.patchTraces(traces, err => {
+        if (err) {
+          const errorMsg = `sendTrace error: ${err.message}`;
+          this.logger.error(errorMsg);
+          reject(errorMsg);
+        } else {
+          const successMsg = 'sendTrace sucessfully';
+          this.logger.debug(successMsg);
+          resolve(successMsg);
+        }
+      });
+    });
+  }
+
+
+
+  /**
+   * Gets the Google Application Credentials from the environment variables,
+   * authenticates the client and calls a method to send the traces data.
+   * @param sendTrace
+   * @param stackdriverTraces
+   */
+  private authorize(stackdriverTraces) {
+    return new Promise((resolve, reject) => {
+      return google.auth.getApplicationDefault(
```

Comment:
This looks like an older version of Google Auth Library. If you add a dependency on `google-auth-library@1.5` (or 1.4+) I believe it comes with a Promise API out-of-the-box.

Reply:
We updated to the last version. It's working with promises now.